README.md
--------------------------------------------
Simple Iptables UI for Gnu/Linux (firewall)
--------------------------------------------

Copyright © 2013 - 2017 by Gianluca Pernigotto aka jeanslack   
Author and Developer: Gianluca Pernigotto   
Mail: <jeanlucperni@gmail.com>   
License: GPL3 (see LICENSE file in the doc folder)

--------------------------------------------

Description:
------- 

cliffwall is a command line user-interface based on iptables for a firewall 
configurations. Can work with any type of network interface with a typical 
home networking 

Inspired by:
[Simple Stateful Firewall](https://wiki.archlinux.org/index.php/Simple_stateful_firewall)

Dependencies:
-------

python >=2.6, iptables


Features:
-------

* Simple stateful firewall rules

* Management for udp/tcp protocol doors rules

* Blocking functions (web sites, network) for a moderate parental control

* ..and other

Building packages:
----
See the INSTALL file in the sources folder for more complete details

* DEBIAN:

Install extra dependencies for build package with distutils:

		# apt-get install python-all python-stdeb fakeroot

Enter in unzipped sources folder and type (with not root):

		python setup.py --command-packages=stdeb.command bdist_deb   

This should create a python-cliffwall.deb in the new deb_dist directory.   
See the INSTALL file for more details and info on how-to build .deb package   
[https://github.com/jeanslack/cliffwall/blob/master/INSTALL](https://github.com/jeanslack/cliffwall/blob/master/INSTALL)

* SLACKWARE:

Require pysetuptools (is not installed for 14.1 and previous version of Slackware:)   
[slackbuild.org](http://slackbuilds.org/repository/14.1/python/pysetuptools/)

Then, download my SlackBuild for compile the binary package:   
[jeanslack slackbuilds](https://github.com/jeanslack/slackbuilds/tree/master/cliffwall)   

See the INSTALL file in the sources folder for more complete details:    [https://github.com/jeanslack/cliffwall/blob/master/INSTALL](https://github.com/jeanslack/cliffwall/blob/master/INSTALL)


--------------------------------------------------------------------------------


CHANGELOG
+--------------------------+
Nov 08  2017 release: 0.14
+--------------------------+

- On Debian 9 stretch my /debian/python-cliffwall.postinst is not has used by 
  stdeb and then is not executed at the post installation of the debian package. 
  This happens if you compile the cliffwall source with distutils and stdeb.
  For now, I have removed the only possibility of using the postinst file 
  and avoid this problem.
  To fix this problem you should do it manually after installation:

+--------------------------+
Nov 07  2017 release: 0.13
+--------------------------+
- Fix error to a path string in the 'cliffwall-init-functions' file. 
  This string regard the 'cliffwall block' command that not was match 
  with the file name:
  /etc/cliffwall/block_network > /etc/cliffwall/block
  
- File and folder placement is now completely relegated to setup.py

+--------------------------+
Febbruary 18  2015
+--------------------------+

Initial release


cliffwall_pack/docstring.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

def menu_doc():
    
    print """\033[1m 
In questo spazio è possibile gestire alcune regole personali
nel firewall e tenerne traccia.\033[0m

inserire numero o lettera opzionali:


------------------------------------------------------
        \033[1mBLOCCO DEI SITI WEB NON GRADITI:\033[0m

\033[32;1m|1|\033[0m   Blocca siti non graditi
\033[32;1m|2|\033[0m   Cancella regole siti non graditi
\033[32;1m|3|\033[0m   Visualizza la lista dei siti bloccati.  
-----------------------------------------------------
        \033[1mIMPOSTAZIONI REGOLE TCP/UDP\033[0m

\033[32;1m|4|\033[0m   Aperture/chiusura delle porte sui protocolli tcp/udp
\033[32;1m|5|\033[0m   Cancella regole precedentemente impostate tcp/udp
\033[32;1m|6|\033[0m   Visualizza la lista regole tcp/udp
-----------------------------------------------------

\033[31;1m|Q|\033[0m   Esci\n"""

def ip_insert_doc():
    
    print '''
Il blocco dei siti non graditi avviene immettendo una parola chiave
che identifica il sito, per esempio: google . 
Per volere limitare il blocco solo a www.google.com ma non a 
www.google.it, si potrebbe restringere il campo immettendo un indirizzo
nella sua interezza, per esempio: www.google.com

Svuotare i dati dal proprio browser rende effettiva la nuova regola. 
'''

def ip_delete_doc():
    
    print '''
La procedura per rimuovere le regole precedentemente sottoscritte av-
viene similmente a quella della loro sottoscrizione, cioè immettendo
la parola o address-name esattamente corrispondente.
Per vedere quali regole sono già state sottoscritte, tornare al menu
e usare l'opzione che visualizza la loro lista. 
'''

def door_connect_doc():
    
    print '''\n
Le porte TCP/UDP sono tutte controllate dalle nuove chain (-N TCP, -N UDP)
impostate in modo predefinito nel firewall. Tuttavia è possibile operare 
per poter aprire/chiudere determinate porte o eliminare delle regole.
Ogni regola è costituita da tre target (drop, accept, reject) che deter-
minano l'azione sulla porta: drop rifiuta e scarta il pacchetto, accept
lo consente e reject lo respinge al mittente.
Qui le porte devono essere indicate con un numero (o una lista di numeri 
delimitati da una virgola fino ad un massimo di 15) o un range (8990:9000), 
dei protocolli e un target.

esempi:

NUMERO: 6890 
LISTA DI NUMERI: 22,25,53,80,433,465,5222,5269,5280,8999:9003
PROTOCOLLO: TCP, UDP o TCP/UDP
TARGET: drop (chiude la porta)
        accept (apre la porta)
        reject ( chiude la porta e respinge al mittente)
        
NOTA: per includere un range o intervallo di porte usa per esempio 6890:6895
'''

def door_delete_doc():

    print '''\n
La procedura per rimuovere le regole precedentemente sottoscritte av-
viene similmente a quella della loro sottoscrizione, cioè immettendo
il numero della porta TCP o UDP che sottointende la regola da eliminare.
Per vedere quali regole sono già state sottoscritte, tornare al menu
e usare l'opzione che visualizza la loro lista.
'''

def tcp_udp_graphic_review(arg):
    """
    Simples graphic for tcp and udp port protocol list
    """
    
    graphic_rules_TCP = '''\033[34;1m
------------------------------------------------     
TCP PORT:                NAME:     TARGETS:
------------------------------------------------\033[0m'''

    graphic_rules_UDP = '''\033[34;1m
------------------------------------------------     
UDP PORT:                NAME:     TARGETS:
------------------------------------------------\033[0m'''
    
    if arg == 'TCP':
        print graphic_rules_TCP
            
    if arg == 'UDP':
        print graphic_rules_UDP
        
        
def blocksites_review_graphic():
    """
    View a simple graphic for websites list
    """
    
    print '''\n\033[1m  Blocked Websites List \033[0m \033[34;1m
------------------------------------------------     
KEY WORD:                         DESCRIPTION:
------------------------------------------------\033[0m'''


cliffwall_pack/conf_parser.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#
#########################################################
# Name: conf_parser.py (module)
# Porpose: make parsing at iface.conf
# Writer: Gianluca Pernigoto <jeanlucperni@gmail.com>
# Copyright: (c) 2013-2015 Gianluca Pernigoto <jeanlucperni@gmail.com>
# license: GPL3
# Version: (V0.13) 13/02/2015
# Rev 
#########################################################

import sys

def parser(pathrule):
    """
    execute a parsing on iface.conf for established which interfaces
    are uncomment on the file and ready to be pass at iptables command.
    """
    iface_conf = '%s/iface.conf' % pathrule
    
    uncomment = []

    fconf = open(iface_conf, 'r')
    list_lines = fconf.readlines()
    fconf.close()

    for line in list_lines:
        if not line.startswith('#'):
            uncomment.append(line.strip()) # strip rimuove tutti \n + spazi e tab
            
    #   rimuovere le stringhe vuote da una lista:
    #cleaned = filter(None, uncomment)
    #   Note however, that filter returns a list in Python 2, but a 
    #   generator in Python 3. You will need to convert it into a list
    #   in Python 3 :
    #cleaned = list(filter(None, uncomment)) 
    #   or use the list comprehension solution:
    cleaned = [x for x in uncomment if x]

    if cleaned == []:
        
        sys.exit('[ERROR]: %s (nessuna interfaccia di rete disponibile)' % 
                                                    iface_conf)
    if len(cleaned) > 1: 
        
        sys.exit('[ERROR]: %s (troppe righe decommentate?)' % iface_conf)
        
    if cleaned[0].startswith('interface_all'):
        
        return 'start', ""
        
    elif cleaned[0].startswith('interface_list='):
        
        slicing = cleaned[0] # the row
        interfaces = slicing[15:] # interfaces only of the row
        
        if slicing[15:] == '':
            
            sys.exit('[ERROR]: %s (la lista su "interface_list=" risulta '
                                            'essere vuota)' % iface_conf)
        return 'start_types', interfaces
            
    elif cleaned[0].startswith('interface_match=eth+'):
        
        return 'start', cleaned[0]
        
    else:
        sys.exit('[ERROR]: %s (Colonne malformate)' % iface_conf) 


cliffwall_pack/load_custom_rules.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#
#########################################################
# Name: load_custom_rules.py (module)
# Porpose: custom rules runner
# Writer: Gianluca Pernigoto <jeanlucperni@gmail.com>
# Copyright: (c) 2013-2015 Gianluca Pernigoto <jeanlucperni@gmail.com>
# license: GPL3
# Version: (V0.13) 10/02/2015
# Rev: (V0.13) 14/02/2015 add status error
#########################################################

"""
Execute services or rules that was are set with 'cliffwall settings'
command. if one of the databases is empty do not run commands, of course 
"""

import subprocess
import shelve

def load_custom(pathDB):
    """
    This function is called by command 'cliffwall start or restart 
    or reload' and during startup the OS .
    Also, here are applied all personal settings established with 
    command 'cliffwall settings'
    """	
    ################ applies block sites rules
    
    diz = shelve.open('%s/sites_block.db'% pathDB,'c') 
    
    if len(diz) <= 0:
        # nessuna regola di blocco siti web impostata 
        # (No rule to block incoming set)
        diz.close()
        
    else:
        # "Attivo blocco IP [in] personale"
        for key in diz.keys():
            
            try:
                subprocess.check_call('iptables -A OUTPUT -p tcp -m string '
                    '--string "%s" --algo kmp -j REJECT' % (key), shell = True)
                
            except subprocess.CalledProcessError:
                
                sys.exit("\033[31;1m[ERROR]\033[0m subprocess.CalledProcessError: "
                            "firewall daemon > 'load_custom:sites_block.db'")
            #print "check_call(iptables -I INPUT -s %s -j DROP " % key
            
        diz.close()
        
    ################ applies tcp rules
    
    diz = shelve.open('%s/doors_rules_TCP.db'% pathDB,'c') 
    
    if len(diz) <= 0:
        # Nessuna regola TCP impostata"
        diz.close()
        
    else:
        # Attivo apertura connessione TCP personale
        val = diz.values()
        tupla = val[0]
        
        for key in diz.keys():
            
            try:
                subprocess.check_call('iptables -A TCP -p tcp --dport %s -j %s' 
                            %(key, tupla[1]), shell = True)
            
            except subprocess.CalledProcessError:
                
                sys.exit("\033[31;1m[ERROR]\033[0m subprocess.CalledProcessError: "
                            "firewall daemon > 'load_custom:doors_rules_TCP.db'")
            #print "check_call(--dport %s -j %s" %(key, tupla[1])
            
        diz.close()
        
    ################ applies udp rules
    
    diz = shelve.open('%s/doors_rules_UDP.db'% pathDB,'c') 
    
    if len(diz) <= 0:
        # Nessuna regola UDP impostata"
        diz.close()
        
    else:
        # Attivo apertura connesione UDP personale
        val = diz.values()
        tupla = val[0]
        
        for key in diz.keys():
            
            try:
                
                subprocess.check_call('iptables -A UDP -p udp --dport '
                        '%s -j %s' % (key, tupla[1]), shell = True)
            
            except subprocess.CalledProcessError:
                
                sys.exit("\033[31;1m[ERROR]\033[0m subprocess.CalledProcessError: "
                            "firewall daemon > 'load_custom:doors_rules_UDP.db'")
            #print "check_call(--dport %s -j %s" %(key, tupla[1])
            
        diz.close()


cliffwall_pack/settings.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
#
#########################################################
# Name: settings.py (module)
# Porpose: Management of custom rules and storage iptables settings
# Writer: Gianluca Pernigoto <jeanlucperni@gmail.com>
# Copyright: (c) 2013-2015 Gianluca Pernigoto <jeanlucperni@gmail.com>
# license: GPL3
# Version: (Ver.0.11) Friday, 29 July 2012
# Rev (v0.13) 04/02/2015
# Rev (v0.13) 09/02/2015
#########################################################

"""
Module called by /etc/rc.d/rc.firewall. It is consists of two
classes for management options and personal firewall rules.
The class Main_Menu is a derivated (child) class instantiated by 
daemon > settings that inherits from the class Rules_Applied . 
The class Rules_Applied is a simple base class for cmd parameters .
"""
from subprocess import call
import sys
import shelve
from docstring import *

####################################################################

class Rules_Applied(object):
    """
    here are applied all the rules regarding the blocking of web pages 
    and the rules regarding the protocols tcp and udp .
    Also, displays UI formated graphic info that show you the
    current settings of tcp/udp and block sites rules
    
    """
    ############################### BLOCK WEB PAGES RULES
    def IP_insert(self, address, description):
        """
        Applied rules for blocked webpages
        """
        diz = shelve.open('%s/sites_block.db'% self.pathDB,'c')
        
        if diz.has_key(address) is False:
            
            diz[address] = description
            
            str_cmd = ('-A OUTPUT -p tcp -m string --string "%s" '
                        '--algo kmp -j REJECT' % (address))
        else:
            
            diz.close()
            sys.exit("\n\033[31;1m Questa regola di 'blocco pagina "
                                "web' risulta già impostata.\033[0m\n")
        
        blocksites_review_graphic()
        
        for key,val in diz.items():
            print("\033[36m  [%s]      --     %s\033[0m" 
                                                % (str(key), str(val)) )
            
        print("\n\033[1m Regola di 'blocco pagina web' impostata nel "
                        "firewall.\033[0m\n")
        diz.close()
        
        return str_cmd
        
    ################################# DELETTING BLOCK SITE RULES
    def IP_delete(self, address):
        """
        Delete rules for blocked webpages
        """
        diz = shelve.open('%s/sites_block.db'% self.pathDB,'c') 
        
        if diz.has_key(address):
            
            del diz[address]
            
            str_cmd = ('-D OUTPUT -p tcp -m string --string "%s" '
                        '--algo kmp -j REJECT' % (address))
        else:
            
            diz.close()
            sys.exit("\n\033[31;1m Regola inesistente\033[0m")
            
            
        blocksites_review_graphic()
        
        for key,val in diz.items():
            print("\033[36m  [%s]      --     %s\033[0m" 
                                            % (str(key), str(val)) )
            
        print ("\n\033[1m Regola di 'blocco pagina web' "
                                    "cancellata nel firewall.\033[0m\n")
        diz.close()
        
        return str_cmd
        
    #####################################  SHOW SITES BLOCKED LIST
    def IP_review(self):
        """
        Shows list with all blocked web sites
        """
        call(['clear'])
        
        diz = shelve.open('%s/sites_block.db'% self.pathDB,'c') 
    
        blocksites_review_graphic()
        
        for key,val in diz.items():
            print("\033[36m  [%s]      --     %s\033[0m" % (str(key), str(val)) )
        print '\n'
        
        diz.close()
        
    ##################################### TCP/UDP PROTOCOLS APPEND:
    def doors_connect(self, doortarget, description, prtcl, PRTCL, target):
        """
        Establish the rules for tcp and udp protocols. Here return values are:
        - doortarget > is door number 
        - description > is the service or a little decription
        - prtcl > is a protocol (udp or tcp)
        - PRTCL > is a protocol ID  
        - target > is the rule (ACCEPT, REJECT or DROP)
        """
        diz = shelve.open('%s/doors_rules_%s.db' % (self.pathDB, PRTCL),'c')
        
        if diz.has_key(doortarget) is True:
            
            nametarg = diz[doortarget]
            
            print ("\n\n\033[31;1m WARNING:\n\033[0m La porta %s %s era già "
                    "stata precedentemente sottoscritta alla regola %s \n"
                    % (PRTCL, doortarget, nametarg[1]))
            
            print ("\033[34;1m DETTAGLI:\n\033[0m %s protocol\n"
                            " Porta: %s\n Servizio: %s\n Target: %s.\n"
                            "\n\033[34;1m COMPORTAMENTO:\n\033[0m"
                            " La regola non ha subito variazioni\n"
                            % (PRTCL, doortarget, nametarg[0], nametarg[1]))
            diz.close()
            
        else:
            
            diz[doortarget] = description,"%s" % target
            
            str_cmd = ('-A %s -p %s -m multiport --dport %s -j %s' 
                            % (PRTCL, prtcl, doortarget, target))
            
            add =  ("\n\033[1m |%s protocol|: Porta '%s' del Servizio "
                    "'%s' è appesa alla regola '%s' nel firewall.\033[0m\n" 
                        %(PRTCL, doortarget, description, target))
            
            tcp_udp_graphic_review(PRTCL)
            
            for key,val in diz.items():
                print("\033[36m  [%s]        --       %s\033[0m" 
                        % (str(key), str(val)) )
            print add
            diz.close()
            
            return str_cmd
            
    ##################################### DELETING MULTIPLE PROTOCOLS:
    def doors_delete(self, doortarget, prtcl, PRTCL):
        """
        deleting previous tcp and udp protocols rules
        """
        diz1 = shelve.open('%s/doors_rules_%s.db'% (self.pathDB, PRTCL),'c')
                        
        if diz1.has_key(doortarget) is True:
            
            nametarg = diz1[doortarget]

            str_cmd = ('-D %s -p %s -m multiport --dport %s -j %s' 
                    % (PRTCL, prtcl, doortarget, nametarg[1]))
                        
            del diz1[doortarget]
            
            tcp_udp_graphic_review(PRTCL)
            
            for key,val in diz1.items():
                print("\033[36m  [%s]        --       %s\033[0m" 
                                % (str(key), str(val)) )
                
            print ("\n\033[1m %s protocol:\n Porta '%s' \n Servizio "
                    "'%s'\n Target '%s'\n Regola cancellata nel firewall."
                    "\033[0m\n" % (PRTCL, doortarget, nametarg[0], nametarg[1]))
            
            diz1.close()
            
            return str_cmd
            
        else:
            print ("\n\033[31;1m WARNING:\033[0m %s protocol: regola "
                        "inesistente nel firewall\n" % PRTCL)
            diz1.close()
            
    ######################################### OVERVIEW PROTOCOLS RULES
    def doors_review(self, PRTCL):
        """
        Shows list with all tcp/udp rules
        """
        #call('clear', shell = True) # this is useless, produce too space
        
        diz = shelve.open('%s/doors_rules_%s.db'% (self.pathDB, PRTCL),'c') 
        
        print '\033[1m\n   LISTA REGOLE PORTE TCP/UDP \033[0m\033[34;1m'
        
        tcp_udp_graphic_review(PRTCL)
        for key,val in diz.items():
            print("\033[36m  [%s]        --       %s\033[0m" 
                            % (str(key), str(val)) )
        print '\n'
        diz.close()
        
class Main_Menu(Rules_Applied):
    """
    Class child that inherit base class Rules_Applied.
    Show a menu with options and then call the method and/or istance
    """
    def __init__(self, pathDB):
        
        self.pathDB = pathDB # pathDB is a path name of database
        
        self.outPut = None # final output state
        
        self.outPut2 = None
        
        call(['clear'])
        
        menu_doc()
        
        choice = raw_input("Attendo un input che corrisponde all'opzione:  > ")
        
        if choice == '1':
            self.IP_insert()
            
        elif choice == '2':
            self.IP_delete()
            
        elif choice == '3':
            self.IP_review() # in the mother class
        
        elif choice == '4':
            self.doors_connect()
            
        elif choice == '5':
            self.doors_delete()
            
        elif choice == '6':
            self.doors_review('TCP') # in the mother class
            self.doors_review('UDP') # in the mother class
            
        elif choice == 'q' or choice == "Q":
            print "cliffwall, Exit"
            sys.exit()
            
        else:
            sys.exit("Errore di immissione, Exit.")
    
    def IP_insert(self):
        """
        Set the variables name and it pass for insert block ip rules
        """
        call(['clear'])
        
        ip_insert_doc()
        
        address_name_target = raw_input("\n\n\033[1m Parola chiave o address-name,"
                            " ['q' per uscire]\033[0m >  ")
        
        if address_name_target == 'q' or address_name_target == 'Q':
            
            sys.exit()
        
        description = raw_input("\n\n\033[1m descrizione del sito.\n\033[0m >  ")
        
        
        self.outPut = Rules_Applied.IP_insert(self, address_name_target, 
                                        description)
    
    def IP_delete(self):
        """
        Set the variables name and it pass for delete block ip rules
        """
        
        call(['clear'])
        
        ip_delete_doc()

        address_name_target = raw_input("\n\n\033[1m Parola chiave o "
                            "address-name, ['q' per uscire]\033[0m >  ")
        
        if address_name_target == 'q' or address_name_target == 'Q':
            
            sys.exit()
            
        self.outPut = Rules_Applied.IP_delete(self, address_name_target)
        
    def doors_connect(self):
        """
        Set the variables name and it pass for new rules at tcp/utp doors
        """
        call(['clear'])
        
        door_connect_doc()
        
        doortarget = raw_input("\n\n\033[1m Port number to be signed in the "
                        "firewall rules ['q/Q' per uscire]\n\033[0m >  ")
        
        if doortarget == 'q' or doortarget == 'Q':
            
            sys.exit()
    
        if str.isdigit(doortarget) is False:
            sys.exit("\n\n\033[31;1mBad option, is not integer\033[0m")
                
        description = raw_input("\n\n\033[1m Description:\n\033[0m >  ")
        
        protocol = raw_input("\n\n\033[1m Protocols: "
                        "\033[32;1m|1|\033[0m(tcp)  \033[32;1m|2|\033[0m(udp)  "
                                            "\033[32;1m|3|\033[0m(tcp/udp) >  ")
        
        target = raw_input("\n\n\033[1m Target "
                    ":\033[32;1m|1|\033[0m(drop) \033[32;1m|2|\033[0m(reject) "
                                "\033[32;1m|3|\033[0m(accept):\033[0m > N°  ")
        if target == "1":
            target = "DROP"

        elif target == "2":
            target = "REJECT" 

        elif target == "3":
            target = "ACCEPT"
        else:
            sys.exit("\n\n\033[31;1mBad option, invalid target\033[0m")
        
        if protocol == "1":
            self.outPut = Rules_Applied.doors_connect(self, doortarget, 
                                            description, 'tcp', 'TCP', target)
            
        elif protocol == "2":
            self.outPut = Rules_Applied.doors_connect(self, doortarget, 
                                            description, 'udp', 'UDP', target)
            
        elif protocol == "3":
            self.outPut = Rules_Applied.doors_connect(self, doortarget, 
                                            description, 'tcp', 'TCP', target)
            
            self.outPut2 = Rules_Applied.doors_connect(self, doortarget, 
                                            description, 'udp', 'UDP', target)
            
        else:
            sys.exit("\n\n\033[31;1mBad option, invalid protocol\033[0m")
            
    def doors_delete(self):
        """
        Set the variables name and it pass for deleting rules at 
        tcp/utp doors
        """
        call(['clear'])
        
        door_delete_doc()
        
        doortarget = raw_input("\n\n\033[1m Numero porta da cancellare dalle "
                        "regole del firewall ['q/Q' per uscire]\n\033[0m >  ")
        
        if doortarget == 'q' or doortarget == 'Q':
            
            sys.exit()
    
        protocol = raw_input("\n\n\033[1m Tipo di Protocollo connessione: "
                        "\033[32;1m|1|\033[0m(tcp)  \033[32;1m|2|\033[0m(udp)  "
                                            "\033[32;1m|3|\033[0m(tcp/udp) >  ")
        
        if protocol == "1":
            self.outPut = Rules_Applied.doors_delete(self, doortarget, 'tcp', 
                                            'TCP')
            
        elif protocol == "2":
            self.outPut = Rules_Applied.doors_delete(self, doortarget, 'udp', 
                                            'UDP')
            
        elif protocol == "3":
            self.outPut = Rules_Applied.doors_delete(self, doortarget, 'tcp', 
                                            'TCP')
            self.outPut2 = Rules_Applied.doors_delete(self, doortarget, 'udp', 
                                            'UDP')
            
        else:
            sys.exit("\n\n\033[31;1mBad option, invalid protocol\033[0m")
